src/shm/pool/multi.rs

```diff
@@ -0,0 +1,192 @@
+}
+
+impl<E: PartialEq + Clone> From<RawPool> for MultiPool<E> {
+    fn from(inner: RawPool) -> Self {
+        Self {
+            buffer_list: Vec::new(),
+            inner
+        }
+    }
+}
+
+impl<I: PartialEq + Clone> MultiPool<I> {
+    /// Resizes the memory pool, notifying the server the pool has changed in size.
+    ///
+    /// The wl_shm protocol only allows the pool to be made bigger. If the new size is smaller than the
+    /// current size of the pool, this function will do nothing.
+    pub fn resize(&mut self, size: usize, conn: &mut ConnectionHandle) -> io::Result<()> {
+        self.inner.resize(size, conn)
+    }
+    fn free(&self, buffer: &wl_buffer::WlBuffer) {
+        if let Some(buffer) = self.buffer_list.iter().find(|b| b.buffer.eq(buffer)) {
+            buffer.free.swap(true, Ordering::Relaxed);
+        }
+    }
+    /// Removes the buffer with the given identifier from the pool and rearranges the others
+    pub fn remove(&mut self, identifier: &I) {
+        if let Some((i, buffer)) =
+            self.buffer_list
+            .iter()
+            .enumerate()
+            .find(|b| b.1.identifier.eq(identifier)) {
+            let mut offset = buffer.offset;
+            self.buffer_list.remove(i);
+            for buffer in &mut self.buffer_list {
+                if buffer.offset > offset {
+                    let l_offset = buffer.offset;
+                    buffer.offset = offset;
+                    offset = l_offset;
+                }
+            }
+        }
+    }
+    /// Returns the buffer associated with the given identifier.
+    /// If it's not possible to use the buffer associated with the identifier, None is returned.
+    pub fn create_buffer<D, U>(
+        &mut self,
+        width: i32,
+        stride: i32,
+        height: i32,
+        identifier: &I,
+        format: wl_shm::Format,
+        udata: U,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+    ) -> Option<(usize, wl_buffer::WlBuffer, &mut [u8])>
+    where
+        D: Dispatch<wl_buffer::WlBuffer, UserData = U> + 'static,
+        U: Send + Sync + 'static,
+    {
+        let mut found_identifier = false;
+        let mut offset = 0;
+        let size = stride * height;
+        let mut index = None;
+        for (i, buffer) in self.buffer_list.iter_mut().enumerate() {
+            if buffer.identifier.eq(identifier) {
+                found_identifier = true;
+                if buffer.free.load(Ordering::Relaxed) {
+                    buffer.size = buffer.size.max(size as usize);
+                    index = Some(i);
+                }
+            } else if offset > buffer.offset {
+                if buffer.free.load(Ordering::Relaxed) {
+                    buffer.offset = offset;
+                } else {
+                    index = None;
+                }
+            } else if found_identifier {
+                break;
+            }
+            offset += buffer.size;
+        }
```

Comment:
Could you add some comments explaining what this loop does? It's really not clear to me why this is changing the offset of others existing buffers when trying to create a buffer.